File: Blog.Core/Infrastructure/FieldSortOrder.cs
namespace Blog.Core.Infrastructure
{
    using Blog.Core.Enums;
    using Blog.Core.Infrastructure.Pagination.Interfaces;
    using System;
    using System.Linq;
    public class FieldSortOrder<T> : ISortCriteria<T> where T : class
    {
        public String Name { get; set; }

        public OrderType Direction { get; set; }

        public FieldSortOrder()
        {
            Direction = OrderType.Ascending;
        }

        public FieldSortOrder(string name, OrderType direction)
        {
            Name = name;
            Direction = direction;
        }

        public IOrderedQueryable<T> ApplyOrdering(IQueryable<T> qry, Boolean useThenBy)
        {
            IOrderedQueryable<T> result;
            var descending = Direction == OrderType.Descending;
            result = !useThenBy ? qry.OrderBy(Name, descending) : qry.ThenBy(Name, descending);
            return result;
        }
    }
}

File: Shared/Blog.FSharp.Core/Infrastructure/FieldSortOrder.fs
namespace Blog.FSharp.Core.Infrastructure

open System.Linq
open Blog.FSharp.Core.Enums
open Blog.FSharp.Core.Infrastructure.Pagination.Interfaces

/// <summary>
/// Field sort order.
/// </summary>
/// <typeparam name="T">Type.</typeparam>
type FieldSortOrder<'T when 'T : not struct>(?name: string, ?direction: OrderType) =

    let mutable name = defaultArg name null
    let mutable direction = defaultArg direction OrderType.Ascending

    /// <summary>
    /// Gets or sets name.
    /// </summary>
    member this.Name
        with get() = name
        and set(value) = name <- value

    /// <summary>
    /// Gets or sets direction.
    /// </summary>
    member this.Direction
        with get() = direction
        and set(value) = direction <- value

    /// <inheritdoc cref="ISortCriteria{T}" />
    interface ISortCriteria<'T> with
        member this.Direction
            with get() = direction
            and set(value) = direction <- value

        member this.ApplyOrdering(qry: IQueryable<'T>, useThenBy: bool) =
            let descending = (direction = OrderType.Descending)
            if useThenBy then
                QueryableExtensions.ThenBy qry name descending
            else
                QueryableExtensions.OrderBy qry name descending